### DISCUSSION TITLE
Stopping event propagation in event subscriber [User Error]

### DISCUSSION THREAD

**Discussion by jdevinemt** [symfony/symfony#44946] [Q&A] [answered]
I am trying to stop event propagation in one of my event subscribers, but it is not working as I expect. When the second registered method is called the event `propagationStopped` property has reverted back to `false`. Here is an example of how my event subscriber is set up:

```
<?php

namespace App\EventSubscriber;

use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpKernel\Event\ControllerEvent;
use Symfony\Component\HttpKernel\KernelEvents;

class ControllerSubscriber implements EventSubscriberInterface
{
	private KernelInterface $kernel;

        public static function getSubscribedEvents(): array
	{
		return [
			KernelEvents::CONTROLLER => [
				['firstMethod', -1],
				['secondMethod', -2],
			],
		];
	}

	public function firstMethod(ControllerEvent $event): void
	{
                // stop propagation if some criteria is met
		$event->stopPropagation();
	}

	public function secondMethod(ControllerEvent $event): void
	{
                // do something else if propagation hasn't been stopped
	}
}
```

If I output the event at the end of `firstMethod` the `propagationStopped` value is `true`, but when I output it at the beginning of `secondMethod` the `propagationStopped` is `false`.

Maybe I'm going about this the wrong way. In my specific case, I'm trying to stop propagation if I detect a route defined by symfony is requested.

**Comment by ahundiak** [comment, marked as answer]
I copied/pasted your subscriber code and added a couple of dump statements.  As expected, the secondMethod is never even called when you stop propagation in the first method.

What exactly were you expecting?

**Reply by jdevinemt** [reply]
Ah, yes. `$event->stopPropagation()` is working as I was expecting. I guess I was interpreting a flaw with my logic with as an unexpected result from the framework. I should have done some more debugging first.

While I'm here though, I'll share the issue I'm trying to resolve. I'm trying to stop event propagation on symfony-related requests, specifically with the web profiler, but ideally it would handle all routes/controllers built into symfony. Here is a revised version of my example class with what I am attempting to do to stop propagation on symfony-related requests.

```
<?php

namespace App\EventSubscriber;

use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpKernel\Event\ControllerEvent;
use Symfony\Component\HttpKernel\KernelEvents;

class ControllerSubscriber implements EventSubscriberInterface
{
	private KernelInterface $kernel;

        public static function getSubscribedEvents(): array
	{
		return [
			KernelEvents::CONTROLLER => [
				['firstMethod', -1],
				['secondMethod', -2],
			],
		];
	}

	public function firstMethod(ControllerEvent $event): void
	{
                // attempting to stop propagation on symony-related requests
		if(substr($event->getRequest()->attributes->get('_route'), 0, 1) === '_') $event->stopPropagation();
	}

	public function secondMethod(ControllerEvent $event): void
	{
                // do something else if propagation hasn't been stopped
               dd('test');
	}
}
```

When testing this on a `_profiler/{token}` page it does stop propagation initially, but then will dump and die `'test'` in the search bar on the left, which seems to come from a second request that does not have a `_route` attribute.

Is there a better way to detect symfony-specific requests? Eventually I won't need to do this, but I'm migrating a legacy application into symfony and have some legacy prepend code I don't want to run in these types of requests for the time being.

**Reply by ahundiak** [reply]
But why?  Stopping the controller event from propagating is not going to stop the request from being processed and do anything really except for messing up any additional controller subscribers.

'legacy prepend code' means what exactly?  

Keep in mind that the profiler requests are completely independent from the normal application requests.  A regular request will return a full html page with (in dev mode) some javascript appended to the end of the body section.  The javascript runs and kicks off it's own request for data.

**Reply by jdevinemt** [reply]
I'm trying to stop propagation for the profiler for a couple reasons. First, I have some authorization that redirects to the login page for requests on my admin subdomain if an admin user is not logged in. I'm not using the symfony security package yet, because this is a legacy app and I haven't migrated that piece of the puzzle yet. Second, I have some legacy code that used to be prepended to all requests via apache config, now I'm handling it in a symfony event subscriber. I don't want this code to run because it runs legacy code that is unnecessary for the profiler like starting a session. The second part is less important to skip. But because of the authorization redirect logic, the web profiler will not work on requests to the admin subomain where an admin user is not logged in.

Here's an example:

```
public function adminAuth(ControllerEvent $event): void
{
     if($event->getRequest()->attributes->get('subdomain') !== 'admin') return;

     $loginPath = '/login';
     if(!User::isAdmin() && $event->getRequest()->getPathInfo() != $loginPath){
	     $event->setController(fn() => new RedirectResponse($loginPath));
	     $event->stopPropagation();
	     return;
     }
}
```

I'm wondering if instead of running this authorization logic based on the subdomain, I should instead run it based on whether the controller to be used resides in the `/src/Controller/Admin` directory. Or possibly based on whether the controller implements some sort of `AdminControllerInterface` interface. But then I'll have to make sure we implement that interface on all of our admin controllers.

Eventually I aim to use the symfony security bundle, but for now I'm just trying to get things to work with legacy code.

**Reply by ahundiak** [reply]
I still don't think you understand that stopping an event propagation does not stop request handling.  Using $event->setController to redirect is all you should need to do in your adminAuth method.  And as far as prepending some code, use your _route check and just don't prepend the code for the profiler paths.

Maybe it would help if you took a look at HttpKernel::handleRaw($request) method to see how kernel events are used.

**Reply by jdevinemt** [reply]
I understand stopping event propagation doesn't stop the request. I just wanted to stop the rest of the events defined in my custom event subscriber from executing on requests for the profiler. You've answered my original question though. I just need to think through a better way to handle my specific case. But in regards to my original question, I'm all set. Thanks for your help!

**Reply by stof** [reply]
stopping propagation does not stop the rest of events in that subscriber. It stops calling all listeners registered for that event after the one calling the method
